File: Exercicios C/Pilhas/PilhaFlexivel/PilhaFlexivel.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula{
    int valor;
    struct Celula *px;
}Celula;

typedef struct Pilha{
    Celula *topo;
}Pilha;

Pilha *criarPilha(){ // "Construtor"
    Pilha *pilha = (Pilha*)malloc(sizeof(Pilha));
    pilha->topo = NULL;
    return pilha;
}

void push(Pilha *pilha, int x){
    Celula *celula = (Celula*)malloc(sizeof(Celula));
    celula->valor = x;
    celula->px = pilha->topo;
    pilha->topo = celula;
}

int pop(Pilha *pilha){
    Celula *celulaRemovida = (Celula*)malloc(sizeof(Celula));

    if(pilha->topo == NULL){
        printf("Pilha Vazia!");
        return -1;
    }

    int valorRemovido = pilha->topo->valor;
    pilha->topo = pilha->topo->px;
    free(celulaRemovida);

    return valorRemovido;

}

void mostrar(Pilha *pilha){
    Celula *i = pilha->topo;

    printf("Pilha: ");
    while(i != NULL){
        printf("%d ", i->valor);
        i = i->px;
    }  

    printf("\n");
}


int main(){
    
    Pilha *pilha = criarPilha();
    int opc;

    do { 
            printf("\n----- MENU -----");
            printf("\nDigite 0 para SAIR");
            printf("\nDigite 1 para INSERIR");
            printf("\nDigite 2 para DELETAR");
            printf("\nDigite 3 para MOSTRAR");
            printf("\nEscolha uma opcao: ");

            scanf("%d", &opc);

            switch(opc) {
                case 1:{
                    printf("\nDigite o valor: ");
                    int valor; 
                    scanf("%d", &valor);
                    push(pilha,valor);
                    break;
            }
                case 2:{
                    int resul = pop(pilha);
                    printf("\nValor removido: %d", resul);
                    break;
            }
                case 3:
                    mostrar(pilha);
                    break;
        }
            
    } while (opc != 0);

    return 0;
    
}

File: Exercicios-JAVA/Pilhas/PilhaFlexivel/PilhaFlexivel.java
import java.util.Scanner;

public class PilhaFlexivel {

    class Celula {
        private int valor;
        private Celula proximo; 

        public Celula(int valor) {
            this.valor = valor;
            this.proximo = null; 
        }

        public int getValor(){
            return valor;
        }
    }

    private Celula topo;

    public PilhaFlexivel() {
        this.topo = null; 
    }



    public void Push(int x){
        Celula novaCelula = new Celula(x); 
        novaCelula.proximo = this.topo;
        this.topo = novaCelula;
    }

    public int Pop(){
        if(topo == null){
            System.out.println("Pilha Vazia");
        }
            int valorRemovido = this.topo.getValor();
            Celula celulaRemovida = this.topo;
            this.topo = this.topo.proximo;
            celulaRemovida.proximo = null;
        

        return valorRemovido;
    }

    public void mostrar(){

        Celula i = this.topo;

        while (i != null) {
            System.out.print(i.getValor() + " ");
            i = i.proximo;
        }

    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        PilhaFlexivel pilha = new PilhaFlexivel();
        int opc;

        do { 
            System.out.println("----- MENU -----");
            System.out.println("Digite 0 para SAIR");
            System.out.println("Digite 1 para INSERIR");
            System.out.println("Digite 2 para DELETAR");
            System.out.println("Digite 3 para MOSTRAR");
            System.out.print("Escolha uma opcao: ");

            opc = scanner.nextInt();

            switch(opc) {
                case 1:{
                    System.out.print("Digite o valor: ");
                    int valor = scanner.nextInt();
                    pilha.Push(valor);
                    break;
                }
                case 2:{
                    int resul = pilha.Pop();
                    System.out.println("Valor removido: " + resul);
                    break;
                }
                case 3:
                    pilha.mostrar();
                    break;
                default:
                    throw new AssertionError();
            }
            
        } while (opc != 0);

    }
    

}